### ISSUE TITLE
Support for Traefik V2

### ISSUE THREAD

**Issue by xanderflood** [jupyterhub/traefik-proxy#97]
It appears that this project isn't currently compatible with Traefik v2 - it makes calls to the `/api/providers/{depends on the provider}/backends` endpoint, which [no longer exists](https://docs.traefik.io/v2.0/operations/api/#endpoints). It aligns roughly with the `/api/http/servers` endpoint, the only important difference being that a single endpoint contains the servers for all providers, so the implementation needs to filter client-side rather 

I need this and am planning to start work on it if I get a chance soon. My hope is that it can be achieved pretty simply:

1. add a new boolean configuration `traefik_api_v2` which defaults to false
2. modify the `_check_for_traefik_endpoint` function in `proxy.py` to branch on this config value - if false, use the existing implementation, and if true, use `/api/http/servers` and look for a server whose name is `{expected_name}@{provider}`

I'm not a pythoner by background but this seems like a manageable job 🤞

**Comment by devnull-mr** [comment]
I am facing the same issue.

I run an external Traefik v2.2 instance and would like to use traefik-proxy instead of the embedded  configurable-http-proxy. In the current state that is not possible as the API has changed with Traefik v2.0, as mentioned above.

I am happy to test any patches before they are being merged into the master branch (and released) !

**Comment by GeorgianaElena** [comment]
Hey @xanderflood and @devnull-mr :wave:

It seems that the [traefik version 2.2](https://github.com/containous/traefik/releases/tag/v2.2.0) released 2 weeks ago, adds support for the etcd and consul backends, so switching to v2 it's actually an option right now :tada: 

> 1. add a new boolean configuration traefik_api_v2 which defaults to false

I'm not sure if it's worth supporting both versions of the traefik api. It might a good idea to switch the entire implementation of traefik-proxy to use traefik v2, especially since it looks like it solves a performance issue with the key-value store provider versions.
@minrk, what do you think about this?

> 2. modify the _check_for_traefik_endpoint function in proxy.py to branch on this config value - if false, use the existing implementation, and if true, use /api/http/servers and look for a server whose name is {expected_name}@{provider}

@xanderflood, Traefik v2 adds a number of breaking changes (not just the api endpoint). This doc https://docs.traefik.io/migration/v1-to-v2/ seems a very good resource about migrating from v1 to v2 if you still want to give it a try. :sunny:

**Comment by xanderflood** [comment]
Thanks a lot for the input @GeorgianaElena! Fortunately I'm not personally migrating Traefik versions, I'm just looking into deploying a new jupyterhub container to an existing Traefik v2.1 installation, so bumping to v2.2 will be easy for me as well.

Come to think of it, I never looked into the details of how the providers themselves changed. Does Traefik v2 store data in etcd in a different format from how Traefik v1 does? The underlying data models of Tv1 and Tv2 are pretty danged different, so it seems likely that is a more involved refactoring than I originally thought.

**Comment by devnull-mr** [comment]
FYI, I started to work on this [here](https://github.com/devnull-mr/traefik-proxy/tree/Traefik_v2)

Only fixed the installer for now to install the latest versions of `traefik/etcd/consul` so the main part of adjusting the proxy to the Traefik v2 API still needs to be done.
Would appreciate some support/collaboration on this.

**Comment by xanderflood** [comment]
@devnull-mr my use case mainly involves the file provider - etcd and consul are both overkill for me, and I'm not really familiar with either of them anyway. Not sure if I'll have time this week but if I do I'll open up a PR against your branch with some work on the toml-file interaction

**Comment by alexleach** [comment]
Hello, I was also hoping to use an existing external Traefik v2 Proxy with a JupyterHub instance. I had a look at your fork @devnull-mr , and looks like no work has been done on it since April last year. Has anyone made any headway with upgrading the traefik-proxy to work with the traefik v2 API?

**Comment by devnull-mr** [comment]
No further progress was made on my fork, that's correct. I also don't intend to continue working on it. So hopefully the main branch will start working on this at some time.

**Comment by alexleach** [comment]
Okay, no problem, but thanks for confirming @devnull-mr .
I have had a go myself now and appear to have got a Traefik file provider working with Traefik v2.4.8. I am unable to test with etcd or consul providers though, so have barely touched those files.
Is anyone able to provide some feedback? Happy to make a PR if that would be a preferred next step?

https://github.com/alexleach/traefik-proxy/commit/41a5ef31aa6304e12d7b74adda77118c32a9a44b

**Comment by alexleach** [comment]
> Okay, no problem, but thanks for confirming @devnull-mr .
> I have had a go myself now and appear to have got a Traefik file provider working with Traefik v2.4.8. I am unable to test with etcd or consul providers though, so have barely touched those files.
> Is anyone able to provide some feedback? Happy to make a PR if that would be a preferred next step?
> 
> https://github.com/alexleach/traefik-proxy/commit/41a5ef31aa6304e12d7b74adda77118c32a9a44b


I received an email shortly after making my commit notifying me that, typically all the tests fail. I will try and get a significant portion to pass and make a follow-up commit to address these failures. Hopefully most can be resolved fairly easily

**Comment by alexleach** [comment]
Hi all,

I've spent quite a bit of time trying to make `jupyter_traefik_proxy` compatible with the Traefik v2 API, and have made a couple of commits referencing this issue (links shown above). Please see the commit text for details on the code changes, but in summary, I've tried to make all tests pass across the (rather excellent and thorough) test suite. This includes tests for etcd3 and consul backends, as well as separate tests using the file provider, to cover both 'yaml' and 'toml' traefik configuration files. I have made relevant changes to the documentation where I thought necessary, as well.

The updates will obviously make some breaking changes to anyone using Traefik v1, so I would like to propose that a new branch be created in the main repo specifically for Traefik v2. I can then open a PR to the new branch, if that sounds sensible? I'm not sure if I can make a PR to a new, non-existent branch, so please could someone with the necessary privileges create a new, v2 branch that I can make a PR to?

As of April this year, Traefik is still releasing updates to its 1.7 branch, and obviously anyone using `jupyterhub_traefik_proxy` currently, must also be using a traefik v1 proxy. Therefore, I think it makes sense to have two branches and make a separate `jupyterhub_traefik_proxy` release for v2, assuming all tests pass and it passes any required quality testing.

The workflow for the latest commit [has just finished](https://github.com/alexleach/traefik-proxy/actions/runs/963832156), and there are still a few errors unfortunately. I'll try and get these final tests to pass in the github workflow today. The [workflow test results](https://github.com/alexleach/traefik-proxy/runs/2893563148?check_suite_focus=true) from the latest commit have the following short summary:-

```
=========================== short test summary info ============================
FAILED tests/test_installer.py::test_version - AssertionError: assert False
FAILED tests/test_proxy.py::test_get_all_routes[no_auth_etcd_proxy] - TypeErr...
FAILED tests/test_proxy.py::test_get_all_routes[auth_etcd_proxy] - TypeError:...
FAILED tests/test_proxy.py::test_get_all_routes[external_etcd_proxy] - TypeEr...
FAILED tests/test_proxy.py::test_get_all_routes[auth_external_etcd_proxy] - T...
FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-zoe] - TypeE...
FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-50fia] - Typ...
FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-\u79c0\u6a39]
FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-~TestJH] - T...
FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-has@] - Type...
FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-zoe] - TypeErro...
FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-50fia] - TypeEr...
FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-\u79c0\u6a39]
FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-~TestJH] - Type...
FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-has@] - TypeErr...
FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-zoe] - Type...
FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-50fia] - Ty...
FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-\u79c0\u6a39]
FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-~TestJH] - ...
FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-has@] - Typ...
FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-zoe]
FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-50fia]
FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-\u79c0\u6a39]
FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-~TestJH]
FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-has@]
========== 25 failed, 186 passed, 475 warnings in 3231.97s (0:53:51) ===========
```

Hopefully I can get these to pass by the end of the day and then if okay with the maintainers, I hope to be ready to make a PR. Please let me know if there is anything else I can do to ease the addition of this new branch, assuming this is something you'd like to incorporate?

Kind regards,
Alex

**Comment by alexleach** [comment]
> Hi all,
> 
> I've spent quite a bit of time trying to make `jupyter_traefik_proxy` compatible with the Traefik v2 API, and have made a couple of commits referencing this issue (links shown above). Please see the commit text for details on the code changes, but in summary, I've tried to make all tests pass across the (rather excellent and thorough) test suite. This includes tests for etcd3 and consul backends, as well as separate tests using the file provider, to cover both 'yaml' and 'toml' traefik configuration files. I have made relevant changes to the documentation where I thought necessary, as well.
> 
> The updates will obviously make some breaking changes to anyone using Traefik v1, so I would like to propose that a new branch be created in the main repo specifically for Traefik v2. I can then open a PR to the new branch, if that sounds sensible? I'm not sure if I can make a PR to a new, non-existent branch, so please could someone with the necessary privileges create a new, v2 branch that I can make a PR to?
> 
> As of April this year, Traefik is still releasing updates to its 1.7 branch, and obviously anyone using `jupyterhub_traefik_proxy` currently, must also be using a traefik v1 proxy. Therefore, I think it makes sense to have two branches and make a separate `jupyterhub_traefik_proxy` release for v2, assuming all tests pass and it passes any required quality testing.
> 
> The workflow for the latest commit [has just finished](https://github.com/alexleach/traefik-proxy/actions/runs/963832156), and there are still a few errors unfortunately. I'll try and get these final tests to pass in the github workflow today. The [workflow test results](https://github.com/alexleach/traefik-proxy/runs/2893563148?check_suite_focus=true) from the latest commit have the following short summary:-
> 
> ```
> =========================== short test summary info ============================
> FAILED tests/test_installer.py::test_version - AssertionError: assert False
> FAILED tests/test_proxy.py::test_get_all_routes[no_auth_etcd_proxy] - TypeErr...
> FAILED tests/test_proxy.py::test_get_all_routes[auth_etcd_proxy] - TypeError:...
> FAILED tests/test_proxy.py::test_get_all_routes[external_etcd_proxy] - TypeEr...
> FAILED tests/test_proxy.py::test_get_all_routes[auth_external_etcd_proxy] - T...
> FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-zoe] - TypeE...
> FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-50fia] - Typ...
> FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-\u79c0\u6a39]
> FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-~TestJH] - T...
> FAILED tests/test_proxy.py::test_check_routes[no_auth_etcd_proxy-has@] - Type...
> FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-zoe] - TypeErro...
> FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-50fia] - TypeEr...
> FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-\u79c0\u6a39]
> FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-~TestJH] - Type...
> FAILED tests/test_proxy.py::test_check_routes[auth_etcd_proxy-has@] - TypeErr...
> FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-zoe] - Type...
> FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-50fia] - Ty...
> FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-\u79c0\u6a39]
> FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-~TestJH] - ...
> FAILED tests/test_proxy.py::test_check_routes[external_etcd_proxy-has@] - Typ...
> FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-zoe]
> FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-50fia]
> FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-\u79c0\u6a39]
> FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-~TestJH]
> FAILED tests/test_proxy.py::test_check_routes[auth_external_etcd_proxy-has@]
> ========== 25 failed, 186 passed, 475 warnings in 3231.97s (0:53:51) ===========
> ```
> 
> Hopefully I can get these to pass by the end of the day and then if okay with the maintainers, I hope to be ready to make a PR. Please let me know if there is anything else I can do to ease the addition of this new branch, assuming this is something you'd like to incorporate?
> 
> Kind regards,
> Alex

Ah, the test failures appear to stem from `jupyterhub/install.py`. I'd forked from 077a6258a950843368b59959bfc762a98fc37b3d, where the latest commit broke the test_warning fixture of tests/test_installer.py. Essentially, where unknown versions of traefik/etcd/consul were requested, the commit made install.py not download them, whereas the original repo issued a warning but downloaded unknown versions anyway. I've reverted the change, and this will hopefully make the next workflow tests pass. (I'd also not updated the default version of etcd, so the default version didn't have a checksum and wouldn't be downloaded. The reasons the etcd tests therefore failed was because there wasn't an etcd on the test server.)

**Comment by alexleach** [comment]
Just to update, all tests now pass on my Traefik v2 branch, commit 508987c1140daf6bb0d898950fdc327fca25eeb9. I'd made a minor mistake when tidying up my code for the previous commit, which is now resolved. I'll try and create a PR to the main repo. Hope that's okay.

I was reading through the https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1162 PR as well, where @minrk mentions ( https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1162#issuecomment-469745419 ) :-

> traefik uses Apache basic or digest auth and there doesn't appear to be a way to put the token into the template in the right format without requiring the user to specify them in their config file after calling `htpasswd` on them.  This is not insurmountable, but it's pretty inconvenient. One hurdle is that these use base64-encoded digests and the sprig digest functions output hex (and base64-encoded hex is not the same as base64-encoded original bytes). The Apache Basic auth spec does define a 'plain' unencrypted format that isn't supported everywhere, and traefik is indeed one of the places that it's not supported.
> 
> I'm trying to avoid forcing the user to generate more secrets to set up one thing, but I may punt and say that you need to call `htpasswd` to paste it into the form traefik wants. That's pretty annoying, though. The alternative is to rely on network policies to protect the proxy (maybe not the best).

I'm quoting this because I've seen Traefik v2 allows for the API passwords to be changed in the dynamic config, either by file or by KV store. I also made a change in proxy.py (ref https://github.com/alexleach/traefik-proxy/commit/faa2832437bf9a2bcd0023a24c34fea150d5d4c1#diff-8dd9ce05cdee91b9579bfc2587c6be429f4f8279513c7e2488b01f17f0f2940aR295), where the api password will get set in the dynamic config, should jupyterhub manage the traefik proxy. This should allow the traefik api password to get set while running, without any downtime.

Hoping that mentioning this may garner some further interest in the traefik v2 proxy :)

**Comment by maulikjs** [comment]
Hey Folks
Just wondering if there has been any update on providing support for traefik v2! I am happy to help in any way I can :)

**Comment by pragyagarg642** [comment]
Hi Team,
Is there any update on this issue for supporting traefik v2?

**Comment by maulikjs** [comment]
Hey @alexleach ,
Sorry I just saw this, I actually switched jobs and have been busy ramping up. Let me go through this over the weekend and get back to you :D

**Comment by alexleach** [comment]
Hi @maulikjs , yes if you could help test some of this, that would be great. I've been traveling a lot recently and am going away again next week until the end of May, so I won't have much of a chance to help with this I'm afraid, but I'll keep an eye out for any comments etc. @GeorgianaElena made some review comments that I started going through before getting tied up with other things. I'm actually using one of my branches in production, I'm sure it might help a lot of others too if it can get merged successfully 🙂

**Comment by minrk** [comment]
closed by #145
